dashboard/test/ui/test_status.js

```diff
@@ -309,45 +309,133 @@ function updateProgressNow() {
+  console.log(`${symbol} ${browser} ${featureKey}`);
+  return metadata;
+}
+
+// Iterate through `tests`, and fetch metadata for each test directly from S3.
+// This function works without a backend, its a fallback in particular to serve
+// our HTML off a static bucket.
+async function fetchMetadataDirectlyFromS3() {
+  // Fetch all test results from S3 in parallel, start them going:
+  const entries = Object.entries(tests).flatMap(([browser, features]) =>
+    Object.keys(features).map(featureKey =>
+      fetchMetadataDirectlyFromS3For(browser, featureKey)
+    )
+  );
+
+  // Now wait for all the parallel fetch() calls to complete, this returns the
+  // same format as `test_logs_controller.rb`, and allows to operate "backendless".
+  return (await Promise.all(entries)).filter(Boolean);
+}
+
+// Fetches test status (pass, fail, etc) from S3 metadata by proxying our call through
+// test_logs_controller.rb. We now support `fetchMetadataDirectlyFromS3`, but this is slower
+// because the browser more heavily restricts the number or parallel requests to
+// the same domain. As a result, we prefer this method.
+async function fetchMetadataFromDashboardAPI(since) {
+  const res = await fetch(`${API_BASEPATH}/${S3_PREFIX}/since/${since}`, {
+    mode: "no-cors",
+  });
+  if (res.ok) {
+    return res.json();
+  } else {
+    throw new Error(`HTTP ${res.status}`);
+  }
+}
+
+// Try to fetch S3 metadata for _output.html files (which contains test status like pass/fail)
+//
+// First we try to fetch it via `test_logs_controller.rb`, this will work when dashboard is running
+// in situations like the test server, or your local dev box.
+//
+// If this fails, we fallback to fetching the S3 metadata directly. This is a little slower,
+// but allows us to support a static-html-only mode, that allows uploading this file to S3, giving
+// us a permalink we can pass back in drone.
+async function fetchMetadata(since) {
+  let metadata
+  try {
+    // Try to fetch from test_logs_controller.rb first:
+    metadata = await fetchMetadataFromDashboardAPI(since)
+  } catch (error) {
+    // Looks like test_logs_controller.rb wasn't there, we're probably running as static HMTML uploaded to S3
+    console.warn(`Failed to fetch metadata from dashboard API, falling back to trying direct S3 access (original error: ${error})`);
+
+    // Maybe this file is hosted on a static S3 bucket? Or dashboard isn't running?
+    // Lets try accessing S3 directly. The bucket IS public after all.
+
+    // However, first we disable auto-refresh, because unlike S3 we can't do
+    // a one-op list that includes the last_modified metadata, so loading this
+    // every 10s would be doing hundreds of HTTP GET calls over and over.
+    disableAutoRefresh();
+
+    metadata = await fetchMetadataDirectlyFromS3()
+    console.log("got this from fetchMetadataDirectlyFromS3", metadata);
+  }
+  console.log(`returning`, metadata)
+  return metadata
+}
+
+async function refresh() {
```

Comment:
`refresh()` has a relatively large diff change because I reworked it as an async function rather than continuing to chain .then .catch promises. This made the "from controller or direct s3" logic much simpler to follow.